### PR TITLE
Improve matching of langfuse traces with resolved bets. Add example

### PR DESCRIPTION
**Pull request by evangriffiths**

### CODE REVIEW

File: prediction_market_agent_tooling/tools/langfuse_client_utils.py [hunk 1]
```diff
@@ -78,25 +129,35 @@ def get_closest_datetime_from_list(
...
-    )
-    # Sanity check - the trace should be after the bet
-    if traces_for_bet[closest_trace_index].timestamp < add_utc_timezone_validator(
-        bet.created_time
-    ):
-        logger.warning(
-            f"No trace for bet on market {bet.market_id} at time {bet.created_time} found"
-        )
+    bet: ResolvedBet, traces: list[ProcessMarketTrace]
+) -> ProcessMarketTrace | None:
```
Comment (evangriffiths) [inline prediction_market_agent_tooling/tools/langfuse_client_utils.py:133]: First use easy-to-filter-on criteria (market id, trade direction+amount) before looking for closest timestamp. Faster and more reliable

File: prediction_market_agent_tooling/tools/langfuse_client_utils.py [hunk 2]
```diff
@@ -1,18 +1,58 @@
...
+    def buy_trade(self) -> Trade:
+        buy_trades = [t for t in self.trades if t.trade_type == TradeType.BUY]
+        if len(buy_trades) == 1:
+            return buy_trades[0]
+        raise ValueError("No buy trade found")
+
+    @staticmethod
+    def from_langfuse_trace(
+        trace: TraceWithDetails,
+    ) -> t.Union["ProcessMarketTrace", None]:
```
Comment (evangriffiths) [inline prediction_market_agent_tooling/tools/langfuse_client_utils.py:35]: If you do `-> "ProcessMarketTrace" | None:` you get a runtime exception: `TypeError: unsupported operand type(s) for |: 'str' and 'NoneType'` 🤔

Comment (kongzii) [reply prediction_market_agent_tooling/tools/langfuse_client_utils.py:35]: Yeah that's unfortunate. Another option is to do `t.Optional["ProcessMarketTrace"]`

File: prediction_market_agent_tooling/tools/langfuse_client_utils.py [hunk 3]
```diff
@@ -47,11 +87,20 @@ def get_traces_for_agent(
     return all_agent_traces
 
 
-def trace_to_omen_agent_market(trace: TraceWithDetails) -> OmenAgentMarket:
-    assert trace.input is not None, "Trace input is None"
-    assert trace.input["args"] is not None, "Trace input args is None"
+def trace_to_omen_agent_market(trace: TraceWithDetails) -> OmenAgentMarket | None:
+    if not hasattr(trace, "input"):
+        breakpoint()
```
Comment (kongzii) [inline prediction_market_agent_tooling/tools/langfuse_client_utils.py:92]: is this intentional to keep here?

Comment (evangriffiths) [reply prediction_market_agent_tooling/tools/langfuse_client_utils.py]: oops, no!

**Review by kongzii — APPROVED** [review]

**Pull request merged**